### ISSUE TITLE
Enhancement: Show diff (and change set details) before updating a stack

### ISSUE THREAD

**Issue by jfhamlin** [Sceptre/sceptre#683]
# Current Behavior

Today, when a user runs a command to update a stack, they see a prompt like this:

```
$ sceptre update prod/vpc
Do you want to update 'prod/vpc' [y/N]:
```

While the user's goal might be to get `prod/vpc` into some new state, this isn't really enough information for her to confidently answer in the affirmative. There are many reasons the answer could be "N," even when trying to make a change:

1. The user hasn't checked out the latest version of the repo containing the sceptre project, and updating will actually revert older changes rather than apply new ones.
2. The user has the latest changes but forgot to add some resource or parameter.
3. The update is destructive or will cause the replacement of some resource, and the user wants to announce or further plan around such a change.
4. ...

# Desired Behavior

For every stack being updated (either a single stack or each stack in a stack group), show the diff of the template, the stack parameters, and if change sets are used, the change set details, before showing a stack update prompt.

For example:
```
$ sceptre update prod/vpc
Changes for 'prod/vpc/vpc.yaml':
Template:
  @@ -27,6 +27,8 @@
         Tags:
         - Key: Name
           Value: !Ref Name
  +      - Key: Test
  +        Value: '123'
     InternetGateway:
       Type: AWS::EC2::InternetGateway
       Properties:

Parameters:
  @@ -2,5 +2,5 @@
   Name: VPC
  -Tenancy: default
  +Tenancy: dedicated

Change Set:
Action Resource                   Replacement?
==============================================
Modify AttachGateway              False
Modify BastionSecurityGroup       Conditional
Modify NATSecurityGroup           Conditional
Modify PrivateSubnetSecurityGroup Conditional
Modify PublicRoute                Conditional
Modify PublicRouteTable           Conditional
Modify VPC                        Conditional

Do you want to update 'prod/vpc' with these changes? [y/N]:
```

If the general proposal sounds reasonable, we can discuss implementation details and any edge/corner cases we should consider.

**Comment by jfhamlin** [comment]
@ngfgrant, let me know your thoughts on this! I'm turning some of the points in https://github.com/cloudreach/sceptre/issues/520#issuecomment-480911084 into enhancement proposals.

**Comment by ngfgrant** [comment]
Hey @jfhamlin

I think this is currently possible with:

`sceptre create`
`sceptre describe` and 
`sceptre execute`

I.e. 

```shell 
$ sceptre create bundle/a/x a-change-set

Do you want to create change set 'a-change-set' for 'bundle/a/x' [y/N]: y
[2019-04-25 11:57:36] - bundle/a/x - Successfully initiated creation of Change Set 'a-change-set'


$ sceptre --output json describe change-set bundle/a/x a-change-set
{
    "ChangeSetName": "a-change-set",
    "StackName": "sceptre-example-bundle-a-x",
    "CreationTime": "2019-04-25 10:57:36.695000+00:00",
    "ExecutionStatus": "AVAILABLE",
    "Status": "CREATE_COMPLETE",
    "Changes": [
        {
            "ResourceChange": {
                "Action": "Modify",
                "LogicalResourceId": "VPC",
                "PhysicalResourceId": "vpc-07b65ecd49f7003d3",
                "ResourceType": "AWS::EC2::VPC",
                "Replacement": "False",
                "Scope": [
                    "Tags"
                ]
            }
        }
    ]
}


$ sceptre execute bundle/a/x a-change-set

Do you want to execute_change_set change set 'a-change-set' for 'bundle/a/x' [y/N]: y
[2019-04-25 11:59:47] - bundle/a/x sceptre-example-bundle-a-x AWS::CloudFormation::Stack UPDATE_IN_PROGRESS User Initiated
[2019-04-25 11:59:55] - bundle/a/x VPC AWS::EC2::VPC UPDATE_IN_PROGRESS
[2019-04-25 11:59:55] - bundle/a/x VPC AWS::EC2::VPC UPDATE_COMPLETE
[2019-04-25 11:59:55] - bundle/a/x sceptre-example-bundle-a-x AWS::CloudFormation::Stack UPDATE_COMPLETE_CLEANUP_IN_PROGRESS
[2019-04-25 12:00:00] - bundle/a/x sceptre-example-bundle-a-x AWS::CloudFormation::Stack UPDATE_COMPLETE
```


The #520 is a discussion about how best to make this process a bit more streamlined. 

As for a `diff` command. I believe there was a command that existed in v1 at some point but was removed by a previous maintainer and I can't see the reasoning. My opinion on `diff` type of stuff is if you are changing production AWS CloudFormation you should be using: 

- some form of version source control such as git,
- some form of peer review process (if possible) and 
- CloudFormation Change Sets.

between these tools and processes making devastating changes should be reduced to a minimum. 


There is something we could do that utilises Drift Detection to first check if the remote CF matches the local CF but again it _should_ if the proper processes are in place and you don't have cowboys that just change production CF in the browser. 

I would be against creating a custom diff command that implements sceptre specific logic as I think this tool should remain a light-weight wrapper around native AWS CF. There are better things for diff and I think it is a mix of process, git, probably drift detection and change sets.

If there is a strong feeling form the community otherwise then definitely happy to reconsider but nervous of scope-creep.

**Comment by jfhamlin** [comment]
Thanks for the reply, @ngfgrant. 

> My opinion on diff type of stuff is if you are changing production AWS CloudFormation you should be using:
>
> - some form of version source control such as git,
> - some form of peer review process (if possible) and
> - CloudFormation Change Sets.
>
> between these tools and processes making devastating changes should be reduced to a minimum.

Even using all of these, if users are updating stacks from the command line at the end of this process, it's possible that:

- They forgot to pull the latest from the `master` branch of the repo with their templates
- Changes in some non-prod stack are live from a WIP branch, and the user doesn't want to clobber them
- The user forgot to examine the change set before running `execute` and didn't realize there was an undesired change or plan (e.g. a resource replacement)

Simply showing what's about to be changed by running the command beats most process for these.

Re your example, it highlights where change sets actually fall short of the mark. Notice that there's no indication _what_ change is being made to the tags, only that the scope of the change is "Tags" and the VPC won't be replaced.

> There is something we could do that utilises Drift Detection to first check if the remote CF matches the local CF but again it should if the proper processes are in place and you don't have cowboys that just change production CF in the browser.

Drift detection would be a welcome addition, but I don't think it helps with the problem I'm describing here, which isn't about drift (where the resources in a stack receive changes from outside of CFN), but about understanding the differences between the deployed stack and what the sceptre project describes before making a change.

> If there is a strong feeling form the community otherwise then definitely happy to reconsider but nervous of scope-creep.

Is there a good venue for a broader (and possibly synchronous) discussion — say, a Slack workspace or channel for sceptre?

Another way of interpreting this proposal is as joining 1, 2, and 4 into a single sceptre command where 2 includes more detail than a change set alone would provide (the existing diff behavior may be appropriate for this). In that way, you can think of this as aligned with the proposal in #520.

![image](https://user-images.githubusercontent.com/375864/56937723-21c8aa80-6ab3-11e9-80fd-c76228fecb2d.png)

As an aside, and to echo the comment I made in #520, it seems to me that this set of commands:

```
- create
- delete
- execute
- launch
- update
```

could be reduced to this set:
```
- update
- delete
```

**Comment by ngfgrant** [comment]
The cli commands we have relating to Change Sets are: 

```
- create
- delete
- execute
- update
- list 
- describe
```

These align with the stages of AWS Change Sets pretty well so will be kept (the image above and in #520 is from AWS). There will be workflows that we should not disrupt by assuming use-cases or joining up commands that are understood to do a specific thing and removing the existing commands.

I can see the benefit of what `diff` could do to show the difference in each line in a file, for sure; although I think this is already possible with `git diff` or in the "Files Changed" section of a PR.

There are lots of different ways people use sceptre - many use it as part of a CI/CD pipeline, for example, and when changes are merged to master a build occurs and the infrastructure is deployed. I think sceptre playing its part - and no more - in the pipeline is important.

For example, Git and GitHub give users superb code history, code diff, a way to work in distributed teams etc etc. So that is why I think that `git + sceptre` deals with the issue discussed. With a `git diff` and a `sceptre describe ...` you can see what was changed and what CF will apply when the change set is eventually executed.

The other points made I think relate to human/process error. Regarding the user forgetting to pull from master before using sceptre to apply the changes, there are more robust ways to handle this e.g. automate the deployment when merging to master or when pushing git tags. In fact, I would strongly recommend using sceptre in this way. You benefit from having a deployment script which could utilise a range of different tools not just sceptre. I would also recommend protecting the master branch to include another review and to be up to date with the base branch, before merging.

You could further limit who can execute Change Sets. In your CI/CD tool you could have a user that can assume a role with ExecuteChangeSet applied. Whereas, your developers could have Read/Write but not Execute. The end result is a flow that follows CloudFormation best practice, completely removes the potential for human error when deploying in any environment.

Re the synchronous chat thing - I am waiting to hear if we can utilise our company Slack for this.

**Comment by jfhamlin** [comment]
I shouldn't have brought up some of my other questions from my comment in #520 here, so I'll leave the command set discussion be and possibly open a separate issue with any questions.

As for this proposal, the essence is that this prompt
```
$ sceptre update prod/vpc
Do you want to update 'prod/vpc' [y/N]:
```
would be far more useful if it showed the difference between the deployed stack (the template and parameters actually in cloudformation) and what is generated locally by sceptre. I love the git and CI/CD processes you've described, but if I'm prompted like this when updating a stack at the command line — the default mode of the `sceptre update`or `sceptre update -c` command — I really just want to know what changes I'm about to make. `git diff` answers a different question, and it takes quite a lot of process and automation and convention to approach a place where I could expect a `git diff` to answer the CFN diff question. And in general, `git` can't come close to answering the question regardless of process. Git can tell me that a change was made to a YAML, JSON, or Python template, but it won't tell me how that change manifests for a particular stack (especially in the case of a Python template).

If reintroducing the standalone `diff` command is a first step we can agree on, I'd be interested in helping with that. I do believe there's an elegant union of Change Set steps 1, 2, and 4 above, and that there's a case to be made for showing diffs when prompting a user whether they want to update a stack.

**Comment by jfhamlin** [comment]
@ngfgrant in case these don't pop up in your inbox otherwise

**Comment by ngfgrant** [comment]
Implementing a meaningful and robust CloudFormation `diff` that is well tested is absolutely non-trivial. An example implementation done well would be the files found in: https://github.com/awslabs/aws-cdk/tree/3270b4727c05156e43101933e22b0b94ca3fdaaf/packages/%40aws-cdk/cloudformation-diff/lib. If your stack allows I would look at using this tool for a proper `diff`.

Anything less than the above implementation I fear could inadvertently mislead the end user. Therefore, I think having the responsibility of understanding the update should sit with the user and I would be remiss if I did not raise concerns about the scope of this tool in this respect.

Incidentally, I think a proper CF diff like you've describe, implemented in Python, would be a very cool standalone open source project.

Hopefully you can see where I am coming from?

**Comment by jfhamlin** [comment]
I definitely get where you're coming from on this, and I'm happy to go the third-party tooling route. Please correct me on this, but for it to work out it looks like sceptre will have to expose a bit more. Let me run through my use case:

Let's say I have two stacks, A and B, where A depends on an output of B. I make a change to both A and B such that the change in A depends on the change in B.

Now, I want to update A and its dependencies (in this case, B), but I want to view the changes — the actual changes in the template and parameters passed to cloudformation and the change set description, not just what `git diff` would show me — before I do.

So far, this seems like a completely reasonable, even basic, use case. Sceptre can't show me the changes, so how would I do this using an external tool? I need access to at least these things:

0. The dependency graph reachable from A
1. Change set for B
2. Current template for B
3. Current parameters for B
4. New template for B
5. New parameters for B
6. [After updating B] Change set for A
7. Current template for A
8. Current parameters for A
9. New template for A
10. [After updating B] New parameters for A

@ngfgrant The CLI provides most of these (some in roundabout ways), but not all (e.g. the dependency graph isn't available). I haven't looked closely at the Sceptre Python API — do you expect these are all available there?

**Comment by jfalkenstein** [comment]
So I hacked together something for this (for my own company uses) that works pretty well, actually, and isn't that significant. It adds dependencies on cfn-flip and deepdiff, but it does the trick!

Anyway, it wouldn't be hard for someone to take this and make this a `sceptre diff` command. The whole thing is only about 100 lines (sans docstrings and tests)

```python
from pathlib import Path
from typing import List, Dict

import click
from boto3 import client
from botocore.exceptions import ClientError
from cfn_flip import load, dump_json, dump_yaml
from deepdiff import DeepDiff
from sceptre.config.reader import ConfigReader
from sceptre.context import SceptreContext
from sceptre.plan.plan import SceptrePlan

current_directory = Path(__file__).parent.absolute()


def get_generated_templates(sceptre_context: SceptreContext,):
    """Uses Sceptre to generate the templates for the passed context. The functionality is NEARLY
    identical to the `sceptre generate` command.

    Args:
        sceptre_context: The context to use for generating the templates

    Returns:
        A dict where the keys are the stack names and the values are the generated template strings
    """
    plan = SceptrePlan(sceptre_context)
    responses = plan.generate()
    return {
        stack.external_name: template
        for stack, template
        in responses.items()
    }


def get_existing_stack_templates(
    generated_stack_names: List[str],
    cloudformation_client
) -> Dict[str, str]:
    """Obtains the stack template strings for the passed stack names

    Args:
        generated_stack_names: The names of the stacks being generated
        cloudformation_client: The boto3 cloudformation client used for

    Returns:
        A dictionary where the keys are stack names and the values are the current template strings
    """
    existing_stack_templates = {}

    for stack_name in generated_stack_names:
        try:
            template = cloudformation_client.get_template(
                StackName=stack_name,
                TemplateStage='Original'
            )
            template_body = template['TemplateBody']
            if not isinstance(template_body, str):
                template_body = dump_json(template_body)
        except ClientError as ex:
            if ex.response['Error']['Code'] == 'ValidationError':
                # Template doesn't exist
                template_body = '{}'
            else:
                raise

        existing_stack_templates[stack_name] = template_body
    return existing_stack_templates


def compare_templates(
    generated_templates: Dict[str, str],
    deployed_templates: Dict[str, str]
) -> Dict[str, DeepDiff]:
    """Compares the generated templates to the deployed templates using DeepDiff.

    Args:
        generated_templates: The templates Sceptre just generated
        deployed_templates: The templates that have already been deployed

    Returns:
        A dictionary where the keys are the stack names and the values are DeepDiff objects
    """
    comparisons = {}
    for stack_name, generated_template in generated_templates.items():
        generated_dict, _ = load(generated_template)
        deployed_template = deployed_templates.get(stack_name, '{}')
        deployed_dict, _ = load(deployed_template)
        comparisons[stack_name] = DeepDiff(
            deployed_dict,
            generated_dict,
            verbose_level=2
        )
    return comparisons


STAR_BAR = '*' * 80
LINE_BAR = '-' * 80


def output_diff_results(stack_name: str, diff: DeepDiff, print_full_template: bool):
    if not len(diff):
        print(f"No difference from deployed stack {stack_name}")
        return
    print(STAR_BAR)
    print(f"Difference detected for stack {stack_name}!")
    print(LINE_BAR)
    print(diff.to_json(indent=4))
    if print_full_template:
        print(LINE_BAR)
        print(f'Deployed template:\n{LINE_BAR}\n{dump_yaml(diff.t1)}')
        print(LINE_BAR)
        print(f'New Template:\n{LINE_BAR}\n{dump_yaml(diff.t2)}')
    print(STAR_BAR)


def get_config_key(sceptre_context: SceptreContext, key: str):
    config_reader = ConfigReader(sceptre_context)
    return config_reader.read(sceptre_context.config_file).get(key)


@click.command()
@click.option(
    '-p',
    '--print-full-templates',
    is_flag=True
)
@click.argument('command-path')
def cli(command_path, print_full_templates):
    sceptre_context = SceptreContext(
        command_path=command_path,
        project_path=str(current_directory),
    )
    region = get_config_key(sceptre_context, 'region')
    cloudformation_client = client('cloudformation', region_name=region)
    generated = get_generated_templates(sceptre_context)
    existing = get_existing_stack_templates(list(generated.keys()), cloudformation_client)
    comparisons = compare_templates(generated, existing)
    for stack, difference in comparisons.items():
        output_diff_results(stack, difference, print_full_templates)


if __name__ == '__main__':
    cli()
```

**Comment by jfalkenstein** [comment]
I actually posted a draft pull request that's way less hacky to create a `sceptre diff` command.

**Comment by mrowlingfox** [comment]
@jfalkenstein  @alexharv074  I figure it's best to have the main discussion here.

Both PR's have value, It's obvious we are trying to achieve very similar outcomes. I think that the biggest gap between the two implementations is the output format. 

If a consensus can be reached regarding the output format I believe this should be simple enough to implement

**Comment by jfalkenstein** [comment]
Fundamentally peaking, my biggest concerns for a "diff" command are this:
* I can produce a text file with the outputs that isn't simply piped from stdout (since that contains cluttered logs and such). This file can be submitted for review.
* It compares the key/value pairs on the templates so it doesn't get all screwed up if someone slides a resource below another one (since resource order doesn't actually matter in a template)
* It compares as much of the stack config vs the deployed config as much as possible so you can see not just how the template has changed (it might not have changed at all) but also what parameters or other configurations have changed.
* It indicates when new resources have been defined or old resources have been removed.

I think a strict text-diff isn't the right approach here, since (like with github) if you move a resource down and another up, it's like "Oh no! Everything has changed!"

**Comment by mrowlingfox** [comment]
I suppose my biggest want from this feature is that it supports (at least) the unidiff format. My team lives and breathes YAML, and that's a format they can digest easily and efficiently. 

Having a way to allow users to dictate the output in another format, potentially using a templating mechanism (j2 anyone 😝 ) might be worth pursuing. 

> Fundamentally peaking, my biggest concerns for a "diff" command are this:
> 
> * I can produce a text file with the outputs that isn't simply piped from stdout (since that contains cluttered logs and such). This file can be submitted for review.

+1 to this

> * It compares the key/value pairs on the templates so it doesn't get all screwed up if someone slides a resource below another one (since resource order doesn't actually matter in a template)

+1 

> * It compares as much of the stack config vs the deployed config as much as possible so you can see not just how the template has changed (it might not have changed at all) but also what parameters or other configurations have changed.

I think this should be fairly obvious when reviewing the configuration being used, but I think it's worthwhile implementing. 

> * It indicates when new resources have been defined or old resources have been removed.

+1

> I think a strict text-diff isn't the right approach here, since (like with github) if you move a resource down and another up, it's like "Oh no! Everything has changed!"

Resource Ordering is the big gotchya here. In the two months we've been using our implementation we haven't experienced any cases where it's bitten us. We exclusively use YAML templates, so that might be why.

One way to avoid this would be to order the resources before doing the diff. This wouldn't cover the scenario where someone changes the order of a list, but I believe that should be flagged as a valid change.

**Comment by jfalkenstein** [comment]
I've reread this whole issue and also the conversation in #520 (both of which seem to have been going on for nearly two years).  I get that using version control can help these changes of configurations be "reviewed" and github provides a very useful "diff" on code/template changes. However, here's where I believe a "diff" command would be invaluable to me (and others) as a part of CI/CD pipelines:
* There are a lot of ways to create "templates", and Sceptre's biggest strength (IMO) is that it can bring all those ways under the same roof. You can create templates using yaml, json, troposphere, cdk, even SAM and Serverless (with some clever hooks or python template files). This makes for an optimal **development** experience with different, distributed teams engaging in infrastructure as code.
* "Templates" can be very dynamic in their construction. Small configurations like sceptre_user_data, jinja configurations, etc... can have large effects on exactly what and how many resources get created/updated and how.
* I work at a FinTech with about 10 different teams. Compliance with standards like PCI, HIPAA, and just general concern around infrastructure security means that all infrastructure changes (especially those with security implications) need to be reviewed by a compliance officers... who cannot be expected to be reviewing pull requests with templates and configurations across a variety of ways to construct templates.
* The bottom line need is this: **We need a way to reduce down changes into something that can be reviewed by compliance to see exactly what is about to be changed in this "deployment" of infrastructure.** Generating a diff file that can be put into a review ticket would be tremendously useful.

Sure, "change sets" can do this _somewhat_, that functionality is limited. For example, we can't `launch` with a change set or create stacks from scratch with change sets. A diff that shows what is about to be launched vs what is currently running would be very useful for getting things through compliance in a timely way.

**Comment by jfalkenstein** [comment]
This has officially been implemented and merged!
